src/platform/nrfconnect/FactoryDataProvider.cpp

```diff
@@ -73,30 +107,94 @@ CHIP_ERROR FactoryDataProvider<FlashFactoryData>::Init()
         return error;
     }
 
-    error = mFlashFactoryData.GetFactoryDataPartition(factoryData, factoryDataSize);
+    return error;
+}
 
-    if (error != CHIP_NO_ERROR)
+#ifdef CONFIG_CHIP_CRYPTO_PSA
+template <class FlashFactoryData>
+CHIP_ERROR FactoryDataProvider<FlashFactoryData>::MoveDACPrivateKeyToSecureStorage(uint8_t * factoryDataPartition,
+                                                                                   size_t factoryDataSize)
+{
+
+    if (!factoryDataPartition || factoryDataSize == 0)
     {
-        ChipLogError(DeviceLayer, "Failed to read factory data partition");
-        return error;
+        return CHIP_ERROR_INVALID_ARGUMENT;
     }
 
-    if (!ParseFactoryData(factoryData, static_cast<uint16_t>(factoryDataSize), &mFactoryData))
+    if (mFactoryData.dac_priv_key.len != kDACPrivateKeyLength)
     {
-        ChipLogError(DeviceLayer, "Failed to parse factory data");
-        return CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+        return CHIP_ERROR_INVALID_LIST_LENGTH;
     }
 
-    // Check if factory data version is correct
-    if (mFactoryData.version != CONFIG_CHIP_FACTORY_DATA_VERSION)
+    uint8_t clearedDACPrivKey[kDACPrivateKeyLength];
+    memset(clearedDACPrivKey, 0x00, sizeof(clearedDACPrivKey));
+
+    // Check if factory data contains DAC private key
+    if (memcmp(mFactoryData.dac_priv_key.data, clearedDACPrivKey, kDACPrivateKeyLength) != 0)
```

Comment:
I think we discussed that first we should check the existence of PSA key, not the content of factory data, right?

Reply:
Maybe I missed that. I will change it then.

Reply:
Hmm. I didn't remember that. :) We still want to clear the FD even if the key is already imported, right?

Reply:
But honestly, what is the difference? What should we do if the key already exists in PSA ITS: just exit the function, or skip importing and remove the dac key if the CONFIG_CHIP_CRYPTO_PSA_MIGRATE_DAC_PRIV_KEY is set?

Reply:
Hmm.. I thought we agreed on sth like:

```
If PSA key does not exist:
   Import Key from factory data

#if CONFIG_CHIP_CRYPTO_PSA_MIGRATE_DAC_PRIV_KEY 
if DAC is not empty:
   Clear the DAC in factory data
#endif
```

But indeed.. now i don't see any edge case with resetting even if we make an if-statement on DAC in factory reset. I believe we discussed one option, when DAC might be partially overwritten due to reset in between.. though i think it's not the issue..